src/Psl/Type/Internal/DictType.php

```diff
@@ -69,25 +83,35 @@ public function coerce(mixed $value): array
     public function assert(mixed $value): array
     {
         if (! is_array($value)) {
-            throw AssertException::withValue($value, $this->toString(), $this->getTrace());
+            throw AssertException::withValue($value, $this->toString());
         }
 
-        $trace = $this->getTrace();
-        $key_type = $this->key_type->withTrace(
-            $trace->withFrame('dict<' . $this->key_type->toString() . ', _>')
-        );
-        $value_type = $this->value_type->withTrace(
-            $trace->withFrame('dict<_, ' . $this->value_type->toString() . '>')
-        );
-
         $result = [];
-
-        /**
-         * @var Tk $k
-         * @var Tv $v
-         */
-        foreach ($value as $k => $v) {
-            $result[$key_type->assert($k)] = $value_type->assert($v);
+        $key_type = $this->key_type;
+        $value_type = $this->value_type;
+
+        $k = $v = null;
+        $trying_key = true;
+
+        try {
+            /**
+             * @var Tk $k
+             * @var Tv $v
+             */
+            foreach ($value as $k => $v) {
+                $trying_key = true;
+                $k_result = $key_type->assert($k);
+                $trying_key = false;
+                $v_result = $value_type->assert($v);
+
+                $result[$k_result] = $v_result;
+            }
+        } catch (AssertException $e) {
+            throw match (true) {
+                $k === null => $e,
```

Comment:
this case is impossible, `$k` would always be non-null, as attemping to iterate over `$value` cannot throw an exception because it is an `array` not an `iterable`.